### ISSUE TITLE
is it safe to use Kokkos_ENABLE_IMPL_CUDA_MALLOC_ASYNC?

### ISSUE THREAD

**Issue by BenWibking** [parthenon-hpc-lab/parthenon#731]
In AMReX, using the CUDA stream-ordered memory allocator improves performance significantly. In Kokkos, I see it is disabled by default due to a bug in UCX: https://github.com/kokkos/kokkos/issues/4228.

Is it safe to build Parthenon+Kokkos with `Kokkos_ENABLE_IMPL_CUDA_MALLOC_ASYNC` enabled? If not, is there a possible workaround to use this for temporary arrays on device that aren't used as MPI buffers?

**Comment by BenWibking** [comment]
Also, is this known to work with any other MPIs, e.g. Cray?

**Comment by pgrete** [comment]
These are good questions that I don't quite have the answer for.
We are aware of the issue with memory alloc/dealloc overhead and tried to work around those where possible.
For example, all comm buffer are allocated in a single View and the individual buffer are subview to that view, as subviews are effectively for free in terms of costs.

As far as I know, no one has tried/tested malloc_async yet, so the answer to "is it safe" is "we don't know".

What kind of temporary arrays do you have in mind? For in kernel temp arrays there's ScratchPadMemory.

**Comment by BenWibking** [comment]
Ah, I was thinking of the arrays used for storing the reconstructed variables at faces, fluxes, etc. I am not that familiar with Kokkos -- how does ScratchPadMemory work under the hood?

**Comment by pgrete** [comment]
I see, this is exactly what we do in AthenaPK, see https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/blob/main/src/hydro/hydro.cpp#L817
Under the hood for cuda it's "shared memory"  and for Kokkos Module 4 of the lecture https://github.com/kokkos/kokkos-tutorials/wiki/Kokkos-Lecture-Series

**Comment by BenWibking** [comment]
> I see, this is exactly what we do in AthenaPK, see https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/blob/main/src/hydro/hydro.cpp#L817 Under the hood for cuda it's "shared memory" and for Kokkos Module 4 of the lecture https://github.com/kokkos/kokkos-tutorials/wiki/Kokkos-Lecture-Series

But `cons.flux` is not in scratchpad memory, or is it?

**Comment by pgrete** [comment]
Correct, we decided to always allocate flux arrays for all three dims.
This uses quite some memory but makes it straightforward to add additional physics that modifies the fluxes on top of the standard fluxes.

**Comment by BenWibking** [comment]
> Correct, we decided to always allocate flux arrays for all three dims. This uses quite some memory but makes it straightforward to add additional physics that modifies the fluxes on top of the standard fluxes.

Can you point me to where this is allocated in AthenaPK?

**Comment by pgrete** [comment]
This is implicitly handled by Parthenon by specifying the "cons" variables with the `WithFluxes` MetaData, see https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/blob/main/src/hydro/hydro.cpp#L543 and https://github.com/lanl/parthenon/blob/develop/docs/interface/Metadata.md#ghost-zones-communication-and-fluxes
The actual allocation happens here: https://github.com/lanl/parthenon/blob/develop/src/interface/variable.cpp#L145

**Comment by BenWibking** [comment]
Thanks.

**Comment by Yurlungur** [comment]
> For example, all comm buffer are allocated in a single View and the individual buffer are subview to that view, as subviews are effectively for free in terms of costs.

This is no longer true. We play this game for fluxes + coarse buffers, but comm buffers are now pooled.

> Also, is this known to work with any other MPIs, e.g. Cray?

I have run with cray-mpich, with only the complications that come with any cray machine.

**Comment by Yurlungur** [comment]
> This is no longer true. We play this game for fluxes + coarse buffers, but comm buffers are now pooled.

In particular, there is a "pool" of comm buffers of the appropriate shape, which we pull from. Only if we run out of free buffers do we allocate more, which are allocated in some fixed batch size.

**Comment by BenWibking** [comment]
> > For example, all comm buffer are allocated in a single View and the individual buffer are subview to that view, as subviews are effectively for free in terms of costs.
> 
> This is no longer true. We play this game for fluxes + coarse buffers, but comm buffers are now pooled.
> 
> > Also, is this known to work with any other MPIs, e.g. Cray?
> 
> I have run with cray-mpich, with only the complications that come with any cray machine.

Just to be clear, does `Kokkos_ENABLE_IMPL_CUDA_MALLOC_ASYNC` + cray-mpich work?

**Comment by Yurlungur** [comment]
> Just to be clear, does Kokkos_ENABLE_IMPL_CUDA_MALLOC_ASYNC + cray-mpich work?

Ah sorry I misunderstood. I haven't tried this. But I assume if you are using `cray-mpich` with the OFI transport layer, not UCX, it should be ok.

**Comment by lroberts36** [comment]
> > This is no longer true. We play this game for fluxes + coarse buffers, but comm buffers are now pooled.
> 
> In particular, there is a "pool" of comm buffers of the appropriate shape, which we pull from. Only if we run out of free buffers do we allocate more, which are allocated in some fixed batch size.

FWIW when the pool runs out, the new comm buffer batch is still created as a single view and then chunked into subviews.

**Comment by BenWibking** [comment]
This may be fixed in UCX 1.14: https://github.com/openucx/ucx/blob/d83ef403646473d76be42cbef03fb38652169f78/NEWS#L49
